linode/kontainer-engine-driver-lke
# ref: Fix release workflow and tidy up CI

## 📝 Description

This pull request makes various CI-related improvements and resolves an issue with the release GitHub Actions workflow.

These changes include:

- Removing all references to Dapper
    - Support for Dapper has already been dropped in this repository (see #38)
- Removing unused scripts from the `scripts` directory
   - Some of these scripts have been moved to Makefile targets; workflows have been updated accordingly
- Switching from golint -> golangci-lint
- Adding a `make fmt` target backed by gofumpt
- Upgrading the Dockerfile
   - Adding a build stage for compiling the driver binary
   - Switching from Ubuntu 18.04 -> Alpine 3.18
- Enabling static linking (`CGO_ENABLED=0`) on all release binaries

Test release: https://github.com/lgarber-akamai/kontainer-engine-driver-lke/releases/tag/v0.0.3
Release workflow run: https://github.com/lgarber-akamai/kontainer-engine-driver-lke/actions/runs/7185618758
E2E test run: https://github.com/lgarber-akamai/kontainer-engine-driver-lke/actions/runs/7185759395/job/19569696573

Related issues:
- #46 

## ✔️ How to Test

All of the following testing instructions assume you have this repository cloned locally.

### Testing the Dockerfile

1. Run the docker-build Makefile target
    - `make docker-build`
2. Run the resulting image with an arbitrary port
    - `docker run --rm -it linode/kontainer-engine-driver-lke:dev 8888`
3. Observe the driver runs as expected
    - You should see an output similar to: `INFO[0000] lke driver 2 up and running on at 127.0.0.1:8888`

### Testing the Linter

1. Run the lint Makefile target
    - `make lint`
2. Observe that golangci-lint is run through Docker and no linting error is raised.

### Validating That Built Binaries Are Statically Linked

1. Run the build CI target.
    - `make build-ci`
2. Use the `file` command to gather information about the built binary.
    - `file bin/kontainer-engine-driver-lke-linux-amd64`
3. Observe the binary is statically linked.
    e.g. 
```
bin/kontainer-engine-driver-lke-linux-amd64: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, Go BuildID=TK8Iz2wRdWZixEM_y_fi/uA949tIKPpzQ6RN7yTGU/yelv8PZW_d27MWGdVdeL/HpwsZiOgErazYbp2m99v, with debug_info, not stripped
```
